pkg/upload/dsym.go

```diff
@@ -0,0 +1,176 @@
+		}
+
+		// If projectRoot is not set explicitly, use uploadInfo to set it
+		if projectRoot == "" {
+			projectRoot = uploadInfo.ProjectRoot
+		}
+
+		// If scheme is set explicitly, check if it exists
+		if scheme != "" {
+			_, err = ios.IsSchemeInPath(path, scheme, projectRoot)
+			if err != nil {
+				return err
+			}
+		} else {
+
+			// Only when the dsym path is not set, try and work out the scheme
+			if dsymPath == "" {
+				// If the scheme is not set explicitly, try to find it
+				scheme, err = ios.GetDefaultScheme(path, projectRoot)
+				if err != nil {
+					return err
+				}
+			}
+		}
+		log.Info("Using scheme: " + utils.DisplayBlankIfEmpty(scheme))
+
+		// If the dsymPath is not fed in via <path>
+		if dsymPath == "" {
+			buildSettings, err = ios.GetXcodeBuildSettings(path, scheme, projectRoot)
+			if err != nil {
+				return err
+			}
+
+			// Build the dsymPath from build settings
+			dsymPath = filepath.Join(buildSettings.ConfigurationBuildDir, buildSettings.DsymName, "Contents", "Resources", "DWARF")
+
+			// Check if dsymPath exists, if not, try alternative path instead
+			err = utils.Path(dsymPath).Validate()
+			if err != nil {
+				log.Info("Could not find dSYM in expected location: " + utils.DisplayBlankIfEmpty(dsymPath))
+				dsymPath = filepath.Join(buildSettings.ConfigurationBuildDir, strings.TrimSuffix(buildSettings.DsymName, ".dSYM"))
+
+				err = utils.Path(dsymPath).Validate()
+				if err != nil {
+					return err
+				}
+				log.Info("Using alternative dSYM path: " + dsymPath)
+			}
+
+		} else {
+			// Use the dsymPath from the command line, check if it's zipped and unzip it if necessary
+			var extractedLocation string
+			extractedLocation, err = utils.ExtractFile(dsymPath, "dsym")
+			if err != nil {
+				log.Warn(utils.DisplayBlankIfEmpty(dsymPath) + " is not a zip file or directory")
+				return err
+			}
+
+			if extractedLocation != "" {
+				log.Info("Unzipped " + dsymPath + " to " + extractedLocation + " for uploading")
+				dsymPath = extractedLocation
+			}
+		}
+
+		dsyms, err = ios.GetDsymsForUpload(dsymPath)
+		if err != nil {
+			return err
+		}
+
+		for _, dsym := range *dsyms {
+			dsymInfo := "(UUID: " + dsym.UUID + ", Name: " + dsym.Name + ", Arch: " + dsym.Arch + ")"
+			log.Info("Uploading dSYM " + dsymInfo)
+			// If the Info.plist path is defined and we still don't know the apiKey or verionName, try to extract them from it
+			if plistPath != "" && (apiKey == "" || versionName == "") {
+				// Read data from the plist
+				plistData, err = ios.GetPlistData(plistPath)
+				if err != nil {
+					return err
+				}
+
+				// Check if the variables are empty, set if they are abd log that we are using setting from the plist file
+				if versionName == "" {
+					versionName = plistData.VersionName
+					log.Info("Using version name from Info.plist: " + utils.DisplayBlankIfEmpty(versionName))
+
+				}
+
+				if apiKey == "" {
+					apiKey = plistData.BugsnagProjectDetails.ApiKey
```

Comment:
I was trying to think of a way that would let the user know that we had no choice but to fetch a blank api-key value from the plist (since it was not defined explicitly by them), so that when the upload server complains about a missing api key they know to go and add it if they were expecting it to be there. Is there a limitation to how useful that might be? I'm aware it creates a little more noise though too

Reply:
It will trigger the validation in `BuildDsymUploadOptions` to say that API key is missing – for consistency with the other platforms (e.g. [here](https://github.com/bugsnag/bugsnag-cli/blob/54914517e7803c0f6d2e6f43bbf28b248b2d0205/pkg/upload/android-proguard.go#L118)), I'd say only log if it's non-empty.